### ISSUE TITLE
[v1.5.0] Release Tracker

### ISSUE THREAD

**Issue by gchanan** [pytorch/pytorch#35009]
The [1.5 release branch](https://github.com/pytorch/pytorch/tree/release/1.5) has been cut!

This issue is for tracking cherry-picks to the release branch.  The criteria for a change to be included is ALL of the following:

1. bug fix or documentation fix/improvement (i.e. NOT a feature)
2. already landed in master or specific to 1.5 (e.g. deprecating a feature in 1.5)
3. worth the risk/reward.  For example, a large change that fixes a tiny bug would not be included.

If you need a change to be cherry-picked to the branch, please open a PR against the [release/1.5](https://github.com/pytorch/pytorch/tree/release/1.5) branch and comment below.  Any details you can provide about the criteria above (e.g. links to master PR, description of the severity of the bug fix) are helpful.  Ultimately we will make an inclusion / no inclusion call and communicate that here.

**NOTE: do not land the PRs yourself.  Our normal tools (ghstack / ghimport, etc.) do not work on the release branch.  Someone else will land your change for you.**

--------

[Current PRs open against the release/1.5](https://github.com/pytorch/pytorch/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+base%3Arelease/1.5)

**Comment by jamesr66a** [comment]
Docs change once it lands https://github.com/pytorch/pytorch/pull/35007

**Comment by VitalyFedyunin** [comment]
#34934 fixes of maxpool (somehow didn't make it into release cut)

**Comment by gchanan** [comment]
@jamesr66a: looks good to go, can you open a PR against the release branch?

**Comment by mcarilli** [comment]
https://github.com/pytorch/pytorch/pull/35011 may be a candidate.  Native automatic mixed precision support is a very important feature for us that I've been writing for over 6 months now, targeting 1.5.  It was approved and merged yesterday, but required a workaround for some unexpected windows failures (see the PR description for the full story).  Build with workaround has passed ci-all and @ezyang is attempting to land now.

**Comment by yf225** [comment]
C++ frontend fixes:

**In Progress**:
- #35022 Fix AdaptiveAvgPool{2,3}d and AdaptiveMaxPool{2,3}d implementation (https://github.com/pytorch/pytorch/pull/35022)
- #35023 Fix Conv and ConvTranspose implementation (https://github.com/pytorch/pytorch/pull/35023)
- #35024 Fix fractional_max_pool3d_with_indices implementation (https://github.com/pytorch/pytorch/pull/35024)
- #35025 Fix F::interpolate and torch::nn::Upsample implementation (https://github.com/pytorch/pytorch/pull/35025)
- #34957 Merge Optimizer and LossClosureOptimizer (https://github.com/pytorch/pytorch/pull/34957)

**Need to cherry-pick to 1.5 branch**:
- #35001 Add xor_convergence test for lbfgs (https://github.com/pytorch/pytorch/pull/35001)

**Comment by jamesr66a** [comment]
https://github.com/pytorch/pytorch/pull/35133

^We'd like to get this in because it aligns torchbind's qualified names with the custom op API, and sets us up to converge the two in the future. Getting this into 1.5 would allow us to save a lot of BC headache in the future

**Comment by zasdfgbnm** [comment]
https://github.com/pytorch/pytorch/pull/35146

^ Docs fix of DP vs DDP on CUDA

**Comment by prajjwal1** [comment]
Leaving it here because this PR fixes a [high priority issue](https://github.com/pytorch/pytorch/issues/33953).  This [PR](https://github.com/pytorch/pytorch/pull/34595) adds Support for Tensor Shape Type Hint.

**Comment by seemethere** [comment]
https://github.com/pytorch/pytorch/pull/35242

^ Script fix to unblock all of CI

**Comment by peterjc123** [comment]
https://github.com/pytorch/pytorch/pull/35275

^ Load torch_global_deps for Windows

**Comment by seemethere** [comment]
https://github.com/pytorch/pytorch/pull/35310

^ Switching default CUDA to 10.2 for wheels

**Comment by ailzhang** [comment]
https://github.com/pytorch/pytorch/pull/35315

Pin XLA CI to use r1.5 release branch.

**Comment by mruberry** [comment]
#35321, required for porting #34794, the latter of which will let users workaround torch.div using either torch.floor_divide or torch.true_divide as appropriate.

**Comment by peterjc123** [comment]
#35368 

^ Load all DLLs in the lib directory for Windows

**Comment by mrshenli** [comment]
Ready patches for PTD:

#35262 enforce rref JIT pickling to be in the scope of rpc calls
#35264 Enforce rref python pickling to be in the scope of RPC call

WIP doc patches for PTD:

#35109 Refactored rpc docs (will need a new PR against release/1.5 after land to master)

cc @zhaojuanmao @rohan-varma

**Comment by mruberry** [comment]
#35390, which ports #34794, which adds a torch.true_divide method variant. This lets users workaround torch.div using either torch.floor_divide or torch.true_divide in all scenarios.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/35349 was opened as a port for https://github.com/pytorch/pytorch/pull/35144.

**Comment by mrshenli** [comment]
#35402 Fix RPC `test_torchscript_functions_not_supported` failure on the release branch. (port for #35283)

cc @xush6528

**Comment by ailzhang** [comment]
[doc]Updating view op list https://github.com/pytorch/pytorch/pull/35403

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/35406, port of #35131.
https://github.com/pytorch/pytorch/pull/35405, port of https://github.com/pytorch/pytorch/pull/35231.
https://github.com/pytorch/pytorch/pull/35407, port of #35053.

**Comment by seemethere** [comment]
https://github.com/pytorch/pytorch/pull/35412

^ To add `future` as a dependency for all of our packages

**Comment by houseroad** [comment]
https://github.com/pytorch/pytorch/pull/35413 bump up ONNX exporter's version (original: https://github.com/pytorch/pytorch/pull/35059)

https://github.com/pytorch/pytorch/pull/35416 fix torch.mm in ONNX exporter (original: https://github.com/pytorch/pytorch/pull/34661)

**Comment by VitalyFedyunin** [comment]
https://github.com/pytorch/pytorch/pull/35434 contains doc strings and documentation updates for channels last

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/35435, port of https://github.com/pytorch/pytorch/pull/35253.
https://github.com/pytorch/pytorch/pull/35436, port of https://github.com/pytorch/pytorch/pull/34822.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/35436, port of https://github.com/pytorch/pytorch/pull/34822.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/35477, port of https://github.com/pytorch/pytorch/pull/35150.

---------------

Waiting on CI.

**Comment by ailzhang** [comment]
#35450 , adding a warning for a known autograd issue in XLA backend.

**Comment by h6197627** [comment]
#35288, fix for Caffe2 mobile compilation

**Comment by h6197627** [comment]
#35573, port of #35288: fix Caffe2 mobile compilation

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/35657, port of https://github.com/pytorch/pytorch/pull/35560: fixes torch.real and torch.imag

**Comment by wenhui-prudencemed** [comment]
> #35340 may be a candidate. Autocasting completes Pytorch's native automatic mixed precision support, which I've been writing for over 6 months now targeting 1.5. It has thorough documentation and test coverage. It was approved and merged already, but reverted for minor fixes.
> 
> Update: PR against master #35102 has been re-merged and stuck the landing so far. #35340 cherry-picks those diffs onto 1.5 (+3 lines of cosmetic docstring changes).
> 
> Feature, sorry :(. It's in nightlies though.

Hi @mcarilli , so is it going to merge into 1.5 or we can only get it from 1.6 nightly buid? (there is no 1.5 nightly build now)

**Comment by peterjc123** [comment]
#35668, port of #35659: [Windows] make torch_cuda's forced link also work for CMake

**Comment by mcarilli** [comment]
@wenhui-prudencemed Looks like autocast won't be merged to 1.5, because it it's a new feature (fails to satisfy 1. in [Greg's original post](https://github.com/pytorch/pytorch/issues/35009#issue-584078008)).

To use autocast in the nightlies or master, see master documentation:
https://pytorch.org/docs/master/amp.html
https://pytorch.org/docs/master/notes/amp_examples.html

**Comment by anjali411** [comment]
https://github.com/pytorch/pytorch/pull/35579 Disabling complex construction because we want to release all complex related features in 1.6

**Comment by mcarilli** [comment]
https://github.com/pytorch/pytorch/pull/35738 relands https://github.com/pytorch/pytorch/pull/35435 with a fix for https://github.com/pytorch/pytorch/issues/35584.

**Comment by mcarilli** [comment]
https://github.com/pytorch/pytorch/pull/35745 adds a warning that 1.5's implementation of automatic mixed precision is incomplete and points people to the master branch/nightly builds.

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/35772 ports disabling torch.imag.

**Comment by yf225** [comment]
In progress:
- #35777 Improve C++ API autograd and indexing docs
- #35190 Refactor C++ API parity test mechanism and turn it on in CI again

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/35786, port of https://github.com/pytorch/pytorch/pull/35784.

**Comment by rohan-varma** [comment]
https://github.com/pytorch/pytorch/pull/35808, port of #35109. Only touches RPC documentation

**Comment by mruberry** [comment]
#35890, port of #35862. Removes potentially confusing integer div warnings.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/35900, port of https://github.com/pytorch/pytorch/pull/35816
https://github.com/pytorch/pytorch/pull/35899, port of https://github.com/pytorch/pytorch/pull/35838.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/36109, port of 36072.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/36079, port of https://github.com/pytorch/pytorch/pull/36074 and https://github.com/pytorch/pytorch/pull/36077.

-----

Merged

**Comment by jamesr66a** [comment]
https://github.com/pytorch/pytorch/pull/36111

Fixes python type annotations for torchbind types in certain situations

**Comment by jlin27** [comment]
https://github.com/pytorch/pytorch/pull/36116

Update docs for 1.5 to remove Python 2 references

**Comment by glaringlee** [comment]
#36141 port from #36095 
fixed issue #36046.

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/36126, port of https://github.com/pytorch/pytorch/pull/36052

**Comment by jlin27** [comment]
https://github.com/pytorch/pytorch/pull/36165

Group libraries in docs table of contents and add link to new PyTorch Elastic page

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/36185, port of 35984.

**Comment by yf225** [comment]
https://github.com/pytorch/pytorch/pull/36245, port of https://github.com/pytorch/pytorch/pull/36161 C++ Adam optimizer - corrected messages for check of default options.

**Comment by yf225** [comment]
https://github.com/pytorch/pytorch/pull/36274, port of https://github.com/pytorch/pytorch/pull/35601 fix is_float_scale_factor warning (python and c++)

**Comment by jlin27** [comment]
https://github.com/pytorch/pytorch/pull/36338

Update docs for 1.5 to remove Python 2 references (Fixes compile error from original PR)

**Comment by bmanga** [comment]
Not sure if this is the right place, but I would consider issue #36378 to be a blocker for release, as for example it prevents building torchvision.

**Comment by mcarilli** [comment]
https://github.com/pytorch/pytorch/pull/36514 minor change to fix formatting errors in the warnings  added by https://github.com/pytorch/pytorch/pull/35745.

**Comment by erip** [comment]
Any possibility of sneaking #35352 into 1.5.0?

**Comment by jinfagang** [comment]
AT_CHECK is gone, what's alternative?

**Comment by mrshenli** [comment]
#36537, port of #36656

Add a warning for Single-Process Multi-GPU DDP

**Comment by mrshenli** [comment]
> AT_CHECK is gone, what's alternative?

@jinfagang TORCH_CHECK is the alternative.

**Comment by gchanan** [comment]
> Any possibility of sneaking #35352 into 1.5.0?

No, this is a feature.

**Comment by peterjc123** [comment]
#36658  Migrate release CI jobs to CircleCI for Windows (v1.5 Release)

**Comment by mrshenli** [comment]
#36732, port of #36675

Doc-only change to make sure existing URLs to RPC docs are still valid.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/36782, port of https://github.com/pytorch/pytorch/pull/36780.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/36762, port of https://github.com/pytorch/pytorch/pull/36760.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/36761, port of https://github.com/pytorch/pytorch/pull/36736.

**Comment by Krovatkin** [comment]
@gchanan 

https://github.com/pytorch/pytorch/pull/36824

Given that we keep running into the issue of incredibly long compilation times, we decided to switch to the simple executor which skips fusion-related passes and analyses. It might affect some users that rely on fusion on GPU to achieve their performance. profile. We will make a note in our release notes for such users to enable the profiling executor.

**Comment by Roshrini** [comment]
@mrshenli  Is this issue getting tracked for release? https://github.com/pytorch/pytorch/issues/36268
PR https://github.com/pytorch/pytorch/pull/36523

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/36839, which fixes https://github.com/pytorch/pytorch/issues/36798.

**Comment by mrshenli** [comment]
disable flaky test #36924

**Comment by mshuaibii** [comment]
pickle support for sparse tensors [#27062](https://github.com/pytorch/pytorch/pull/27062)

**Comment by ailzhang** [comment]
https://github.com/pytorch/pytorch/pull/36927 fixes xla job failure on release/1.5 branch.

**Comment by Krovatkin** [comment]
@gchanan https://github.com/pytorch/pytorch/pull/36933 should fix the test failures

**Comment by mrshenli** [comment]
#36947, port for #36948 

Fixes build error when `USE_DISTRIBUTED=0`. This does  **not** need to be added into our release binaries as they are all built with `USE_DISTRIBUTED=1`, but needs to be landed into the `release/1.5` source branch in case users would like to build using `USE_DISTRIBUTED=0`.

**Comment by gchanan** [comment]
> @mrshenli  Is this issue getting tracked for release? https://github.com/pytorch/pytorch/issues/36268
PR https://github.com/pytorch/pytorch/pull/36523

No, we weren't confident in the fix so close to the release.
